source/slang/slang-emit-spirv.cpp

```diff
@@ -807,101 +918,104 @@ struct SPIRVEmitContext
-    // opcode, including whether a type <id> or result <id> is
-    // required.
-
-    SpvInst* emitInst(SpvInstParent* parent, IRInst* irInst, SpvOp opcode)
+    // This function performs no checks that it is actually being used
+    // correctly with respect to the SPIR-V rules for each opcode. As such, a
+    // more type safe function for each opcode is included in
+    // 'slang-emit-spirv-ops.h', and available in this class. You are
+    // encouraged to use these instead.
+    //
+    template<typename... Operands>
+    SpvInst* emitInst(SpvInstParent* parent, IRInst* irInst, SpvOp opcode, const Operands& ...ops)
     {
         InstConstructScope scopeInst(this, opcode, irInst);
         SpvInst* spvInst = scopeInst;
+        (emitOperand(ops), ...);
         parent->addInst(spvInst);
         return spvInst;
     }
 
-    template<typename A>
-    SpvInst* emitInst(SpvInstParent* parent, IRInst* irInst, SpvOp opcode, A const& a)
+    template<typename OperandEmitFunc>
+    SpvInst* emitInstCustomOperandFunc(SpvInstParent* parent, IRInst* irInst, SpvOp opcode, const OperandEmitFunc& f)
     {
         InstConstructScope scopeInst(this, opcode, irInst);
         SpvInst* spvInst = scopeInst;
-        emitOperand(a);
+        f();
         parent->addInst(spvInst);
         return spvInst;
     }
 
-    template<typename A, typename B>
-    SpvInst* emitInst(SpvInstParent* parent, IRInst* irInst, SpvOp opcode, A const& a, B const& b)
+    // Emits a SPV Inst with deduplication
+    // This is used where our IR doesn't guarantee uniqueness but SPIR-V
+    // requires it
+    template<typename... Operands>
+    SpvInst* emitInstMemoized(
+        SpvInstParent* parent,
+        IRInst* irInst,
+        SpvOp opcode,
+        // We take the resultId here explicitly here to make sure we don't try
+        // and memoize its value.
+        ResultIDToken resultId,
+        const Operands& ...ops
+    )
     {
-        InstConstructScope scopeInst(this, opcode, irInst);
-        SpvInst* spvInst = scopeInst;
-        emitOperand(a);
-        emitOperand(b);
-        parent->addInst(spvInst);
-        return spvInst;
-    }
+        List<SpvWord> ourOperands;
+        {
+            auto scopePeek = OperandMemoizeScope(this);
+            (emitOperand(ops), ...);
+            // Steal our operands back, so we don't have to calculate them
+            // again
+            ourOperands = std::move(m_operandStack);
+        }
 
-    template<typename A, typename B, typename C>
-    SpvInst* emitInst(SpvInstParent* parent, IRInst* irInst, SpvOp opcode, A const& a, B const& b, C const& c)
-    {
-        InstConstructScope scopeInst(this, opcode, irInst);
-        SpvInst* spvInst = scopeInst;
-        emitOperand(a);
-        emitOperand(b);
-        emitOperand(c);
-        parent->addInst(spvInst);
-        return spvInst;
-    }
+        // Hash the whole global stack and opcode
+        SpvTypeInstKey key;
```

Comment:
For better performance, we might want to use the spvInst itself as the storage for the opcode+operands instead of using a List as dictionary key, similar to how `IRInstKey` and `ValKey` are implemented.
```
struct SpvTypeInstKey
{
      SpvInst* inst;
      bool operator == (SpvTypeInstDesc desc) { ... }
}

struct SpvTypeInstDesc
{
       SpvOp opCode;
       ArrayView<SpvWord> operands;
}
```

Reply:
Good idea, I wonder if we could also use our IRInst as an initial guess, I don't think we can generate a different SPIR-V instruction for the same IRInst ever